static/js/lib/learning_resources.js

```diff
@@ -27,24 +34,131 @@ export const availabilityLabel = (availability: ?string) => {
   }
 }
 
+export const parseDateFilter = (filter: string) => {
+  const format = /(now)(\+)?(\d+)?([Md])?/
+  const match = format.exec(filter)
+  if (match) {
+    let dt = moment()
+    if (match[3] && match[4]) {
+      dt = dt.add(match[3], match[4] === "d" ? "days" : "months")
+    }
+    return dt
+  }
+}
+
+export const inDateRanges = (run: CourseRun, availabilities: Array<string>) => {
+  let inRange = false
+  availabilities.forEach(availability => {
+    if (AVAILABILITY_MAPPING[availability]) {
+      const from = parseDateFilter(
+        AVAILABILITY_MAPPING[availability].filter.from
+      )
+      const to = parseDateFilter(AVAILABILITY_MAPPING[availability].filter.to)
+      const startDate = run.best_start_date
+        ? moment(run.best_start_date, dateFormat)
+        : null
+      if (
+        ((isNil(startDate) && availability === AVAILABLE_NOW) ||
+          isNil(from) ||
+          // $FlowFixMe: okay to compare moments
+          startDate >= from) &&
+        ((isNil(startDate) && availability === AVAILABLE_NOW) ||
+          isNil(to) ||
+          // $FlowFixMe: okay to compare moments
+          startDate <= to)
+      ) {
+        inRange = true
+      }
+    }
+  })
+  return inRange
+}
+
+export const bestRunLabel = (run: ?CourseRun) => {
+  if (!run) {
+    return AVAILABILITY_MAPPING[AVAILABLE_NOW].label
+  }
+  for (const range in AVAILABILITY_MAPPING) {
+    if (inDateRanges(run, [range])) {
+      return AVAILABILITY_MAPPING[range].label
+    }
+  }
+}
+
+export const compareRuns = (firstRun: CourseRun, secondRun: CourseRun) =>
+  moment(firstRun.best_start_date || defaultStartDate, dateFormat).diff(
+    moment(secondRun.best_start_date || defaultStartDate, dateFormat),
```

Comment:
a general suggestion for a lot of this code: this call is repeated a whole bunch of times:

```js
moment(run.best_start_date || defaultStartDate, dateFormat)
```

along with a closely-related call for the end date. The code overall might be easier to read and follow if you write two little helpers, like `runStartDate` and `runEndDate`, that do this for you in one place.

Reply:
✔️